src/dev/flang/ast/NumLiteral.java

```diff
@@ -423,6 +426,63 @@ AbstractType typeForInferencing()
   }
 
 
+  /**
+   * typeForUnion returns the type of this expression or null if the type is
+   * still unknown, i.e., before or during type resolution.  This is redefined
+   * by sub-classes of Expr to provide type information.
+   *
+   * @return this Expr's type or null if not known.
+   */
+  @Override
+  AbstractType typeForUnion()
+  {
+    return _type;
+  }
+
+
+  /**
+   * During type inference: Inform this expression that it is
+   * expected to result in the given type.
+   *
+   * @param t the expected type.
+   */
+  @Override
+  protected void propagateExpectedType(AbstractType t)
+  {
+    if (PRECONDITIONS) require
+      (_type == null || extractNumericType(t) == null || _type.compareTo(extractNumericType(t)) == 0);
+
+    _type = extractNumericType(t);
+  }
+
+
+  private AbstractType extractNumericType(AbstractType t)
+  {
+    var numericTypes = new TreeSet<AbstractType>(
```

Comment:
This should better be a field in `Types.resolved` to avoid re-creation of this `TreeSet` for every call to `propagateExpectedType`.